// posthog-js/lib/src/utils/property-utils.js
import { jsonStringify } from '../request';
import { isMatchingRegex } from './regex-utils';
export function getPersonPropertiesHash(distinct_id, userPropertiesToSet, userPropertiesToSetOnce) {
    return jsonStringify({ distinct_id: distinct_id, userPropertiesToSet: userPropertiesToSet, userPropertiesToSetOnce: userPropertiesToSetOnce });
}
export var propertyComparisons = {
    exact: function (targets, values) { return values.some(function (value) { return targets.some(function (target) { return value === target; }); }); },
    is_not: function (targets, values) { return values.every(function (value) { return targets.every(function (target) { return value !== target; }); }); },
    regex: function (targets, values) { return values.some(function (value) { return targets.some(function (target) { return isMatchingRegex(value, target); }); }); },
    not_regex: function (targets, values) { return values.every(function (value) { return targets.every(function (target) { return !isMatchingRegex(value, target); }); }); },
    icontains: function (targets, values) {
        return values.map(toLowerCase).some(function (value) { return targets.map(toLowerCase).some(function (target) { return value.includes(target); }); });
    },
    not_icontains: function (targets, values) {
        return values.map(toLowerCase).every(function (value) { return targets.map(toLowerCase).every(function (target) { return !value.includes(target); }); });
    },
};
var toLowerCase = function (v) { return v.toLowerCase(); };
