// posthog-js/lib/src/consent.js
import { find } from './utils';
import { assignableWindow, navigator } from './utils/globals';
import { cookieStore, localStore } from './storage';
import { includes } from './utils/string-utils';
var OPT_OUT_PREFIX = '__ph_opt_in_out_';
export var ConsentStatus;
(function (ConsentStatus) {
    ConsentStatus[ConsentStatus["PENDING"] = -1] = "PENDING";
    ConsentStatus[ConsentStatus["DENIED"] = 0] = "DENIED";
    ConsentStatus[ConsentStatus["GRANTED"] = 1] = "GRANTED";
})(ConsentStatus || (ConsentStatus = {}));
/**
 * ConsentManager provides tools for managing user consent as configured by the application.
 */
var ConsentManager = /** @class */ (function () {
    function ConsentManager(_instance) {
        this._instance = _instance;
    }
    Object.defineProperty(ConsentManager.prototype, "_config", {
        get: function () {
            return this._instance.config;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ConsentManager.prototype, "consent", {
        get: function () {
            if (this._getDnt()) {
                return ConsentStatus.DENIED;
            }
            return this._storedConsent;
        },
        enumerable: false,
        configurable: true
    });
    ConsentManager.prototype.isOptedOut = function () {
        return (this.consent === ConsentStatus.DENIED ||
            (this.consent === ConsentStatus.PENDING && this._config.opt_out_capturing_by_default));
    };
    ConsentManager.prototype.isOptedIn = function () {
        return !this.isOptedOut();
    };
    ConsentManager.prototype.optInOut = function (isOptedIn) {
        this._storage._set(this._storageKey, isOptedIn ? 1 : 0, this._config.cookie_expiration, this._config.cross_subdomain_cookie, this._config.secure_cookie);
    };
    ConsentManager.prototype.reset = function () {
        this._storage._remove(this._storageKey, this._config.cross_subdomain_cookie);
    };
    Object.defineProperty(ConsentManager.prototype, "_storageKey", {
        get: function () {
            var _a = this._instance.config, token = _a.token, opt_out_capturing_cookie_prefix = _a.opt_out_capturing_cookie_prefix;
            return (opt_out_capturing_cookie_prefix || OPT_OUT_PREFIX) + token;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ConsentManager.prototype, "_storedConsent", {
        get: function () {
            var value = this._storage._get(this._storageKey);
            return value === '1' ? ConsentStatus.GRANTED : value === '0' ? ConsentStatus.DENIED : ConsentStatus.PENDING;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ConsentManager.prototype, "_storage", {
        get: function () {
            if (!this._persistentStore) {
                var persistenceType = this._config.opt_out_capturing_persistence_type;
                this._persistentStore = persistenceType === 'localStorage' ? localStore : cookieStore;
                var otherStorage = persistenceType === 'localStorage' ? cookieStore : localStore;
                if (otherStorage._get(this._storageKey)) {
                    if (!this._persistentStore._get(this._storageKey)) {
                        // This indicates we have moved to a new storage format so we migrate the value over
                        this.optInOut(otherStorage._get(this._storageKey) === '1');
                    }
                    otherStorage._remove(this._storageKey, this._config.cross_subdomain_cookie);
                }
            }
            return this._persistentStore;
        },
        enumerable: false,
        configurable: true
    });
    ConsentManager.prototype._getDnt = function () {
        if (!this._config.respect_dnt) {
            return false;
        }
        return !!find([
            navigator === null || navigator === void 0 ? void 0 : navigator.doNotTrack, // standard
            navigator === null || navigator === void 0 ? void 0 : navigator['msDoNotTrack'],
            assignableWindow['doNotTrack'],
        ], function (dntValue) {
            return includes([true, 1, '1', 'yes'], dntValue);
        });
    };
    return ConsentManager;
}());
export { ConsentManager };
